File: p09ab.clj
;;;;    Advent of Code 2023 - Day 9 part One and Two
;;;;    https://adventofcode.com/2023/day/9
;;;;
;;;;    Solutions in Clojure
;;;;    (Ter leering ende vermaeck...)
;;;;
;;;;    The sum of these extrapolated future values: 1939607039
;;;;    The sum of these extrapolated past values:         1041
;;;;
;;;;    (cl) by Arno Jacobs, 2023-12-09
;;;;    

(ns p09ab
  (:require [clojure.string :as str]))

;;;; First - read the data-set
(defn parse-line [line]
  (mapv parse-long (str/split (str/trim line) #" ")))

(defn get-lines [file]
  (into []
   (for [line (str/split-lines (slurp file))]
     (parse-line line))))

(def values (get-lines "data/inputDay09_2023.txt"))


(defn delta-lists [long-list short-list]
  (if (empty? short-list)
    nil
    (cons (- (first short-list) (first long-list)) 
          (delta-lists (rest long-list) (rest short-list)))))

(defn deltas [values]
  (delta-lists values (rest values)))

;;; Part 1

(defn all-future-delta-zeros [border-values values]
  (let [new-border-values (cons (last values) border-values)]
  (if (every? zero? values)
    new-border-values
    (all-future-delta-zeros new-border-values (deltas values)))))

(defn calculate-to-future [values]
  (reduce + (all-future-delta-zeros nil values)))

(defn work-future [values]
  (reduce + (map calculate-to-future values)))

;;; Part 2

(defn all-past-delta-zeros [border-values values]
  (let [new-border-values (cons (first values) border-values)]
    (if (every? zero? values)
      new-border-values
      (all-past-delta-zeros new-border-values (deltas values)))))

(defn extrapolate-past [sign current-value values]
  (if (empty? values) 
    (* sign current-value)
    (extrapolate-past (- sign) (- (first values) current-value) (rest values))))
                      
(defn calculate-to-past [values]
  (extrapolate-past -1 0 (reverse (all-past-delta-zeros nil values))))

(defn work-past [values]
  (reduce + (map calculate-to-past values)))

(defn program []
  (println "Advent of Code 2023 - day 9  (Clojure)")
  (print   "The sum of these extrapolated future values: ")
  (println (work-future values))
  (print   "The sum of these extrapolated past   values:       ")
  (println (work-past values))
  (println "0K.\n"))

;; Run in terminal via: clojure -M p08ab.clj
(program)

File: p09ab.hs
-- Advent of Code 2023 - Day 9 part One and Two
-- Solutions in Haskell
-- (Ter leering ende vermaeck...)
--
-- The sum of these extrapolated future values: 1939607039
-- The sum of these extrapolated past values:         1041

--
-- (cl) by Arno Jacobs, 2023-12-09

module AoC2023d09ab where

import Data.List.Split

-- Some initials
filename :: String
filename = "data/inputDay09_2023.txt"

data Time = Past | Future deriving (Eq,Show)

sSpace = " " :: String

getInts :: String -> [Int]
getInts = map read . splitOn sSpace

-- Part 1

deltas :: [Int] -> [Int]
deltas xs = map (\p -> snd p - fst p) $ zip xs (tail xs)
        
-- Keep every first or last value of the list at each delta function
-- dieection is either Past or Future
allDeltasZeros :: Time -> [Int] -> [Int] -> [Int]
allDeltasZeros direction savedValues values 
    | allZeros && direction == Past     = newFirstValues
    | allZeros && direction == Future   = newFirstValues
    | direction == Past                 = allDeltasZeros direction newFirstValues newValues 
    | otherwise                         = allDeltasZeros direction newLastValues  newValues
        where
            newValues       = deltas values
            newFirstValues  = [head values] ++ savedValues
            newLastValues   = [last values] ++ savedValues
            allZeros        = [] == filter (/=0) newValues 

calculateToFuture :: [Int] -> Int
calculateToFuture = sum . (allDeltasZeros Future [])

workToFuture :: [[Int]] -> Int
workToFuture = sum . map calculateToFuture

-- Part 2

-- "Je gaat het pas zien als je het door hebt." J.C.
--
calculateToPast :: [Int] -> Int
calculateToPast = subtracts . (allDeltasZeros Past [])
    where
        subtracts = foldl subtract 0

workToPast :: [[Int]] -> Int
workToPast = sum . map calculateToPast

main :: IO ()
main = do   putStrLn "Advent of Code 2023 - day 9  (Haskell)"
            histories <- map getInts <$> lines <$> readFile filename
            putStr   "The sum of these extrapolated future values: "
            print $ workToFuture histories
            putStr   "The sum of these extrapolated past values:         "
            print $ workToPast histories
            putStrLn "0K.\n"

